# KNN - Predict the Price of Bangalore House



In [1]:
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
import math

from sklearn.model_selection import train_test_split
from sklearn.neighbors import KNeighborsRegressor
from sklearn.metrics import r2_score
from sklearn.metrics import accuracy_score

In [2]:
data =pd.read_csv("C:/Users/soura/OneDrive/Desktop/Vaibhav DS/vibs project/Capstone project 1/Capstone Project  - 1 KNN Prj_5&Prj_6/KNN Assignment/bangalore house price prediction OHE-data.csv")

In [3]:
print(data.head(2))

   bath  balcony  price  total_sqft_int  bhk  price_per_sqft  \
0   3.0      2.0  150.0          1672.0    3     8971.291866   
1   3.0      3.0  149.0          1750.0    3     8514.285714   

   area_typeSuper built-up  Area  area_typeBuilt-up  Area  \
0                              1                        0   
1                              0                        1   

   area_typePlot  Area  availability_Ready To Move  ...  \
0                    0                           1  ...   
1                    0                           1  ...   

   location_Kalena Agrahara  location_Horamavu Agara  location_Vidyaranyapura  \
0                         0                        0                        0   
1                         0                        0                        0   

   location_BTM 2nd Stage  location_Hebbal Kempapura  location_Hosur Road  \
0                       0                          0                    0   
1                       0                      

In [4]:
data = pd.DataFrame(data)
data

bath  balcony   price  total_sqft_int  bhk  price_per_sqft  \
0      3.0      2.0  150.00          1672.0    3     8971.291866   
1      3.0      3.0  149.00          1750.0    3     8514.285714   
2      3.0      2.0  150.00          1750.0    3     8571.428571   
3      2.0      2.0   40.00          1250.0    2     3200.000000   
4      2.0      2.0   83.00          1200.0    2     6916.666667   
...    ...      ...     ...             ...  ...             ...   
7115   3.0      2.0  325.00          2900.0    3    11206.896552   
7116   3.0      1.0   84.83          1780.0    3     4765.730337   
7117   2.0      1.0   48.00           880.0    2     5454.545455   
7118   2.0      1.0   55.00          1000.0    2     5500.000000   
7119   2.0      1.0   78.00          1400.0    3     5571.428571   

      area_typeSuper built-up  Area  area_typeBuilt-up  Area  \
0                                 1                        0   
1                                 0                        1   
2                                 1                        0   
3                                 1                        0   
4                                 0                        0   
...                             ...                      ...   
7115                              1                        0   
7116                              1                        0   
7117                              0                        0   
7118                              0                        0   
7119                              0                        0   

      area_typePlot  Area  availability_Ready To Move  ...  \
0                       0                           1  ...   
1                       0                           1  ...   
2                       0                           1  ...   
3                       0                           1  ...   
4                       1                           1  ...   
...                   ...                         ...  ...   
7115                    0                           1  ...   
7116                    0                           0  ...   
7117                    1                           1  ...   
7118                    1                           0  ...   
7119                    1                           0  ...   

      location_Kalena Agrahara  location_Horamavu Agara  \
0                            0                        0   
1                            0                        0   
2                            0                        0   
3                            0                        0   
4                            0                        0   
...                        ...                      ...   
7115                         0                        0   
7116                         0                        0   
7117                         0                        0   
7118                         0                        0   
7119                         0                        0   

      location_Vidyaranyapura  location_BTM 2nd Stage  \
0                           0                       0   
1                           0                       0   
2                           0                       0   
3                           0                       0   
4                           0                       0   
...                       ...                     ...   
7115                        0                       0   
7116                        0                       0   
7117                        0                       0   
7118                        0                       0   
7119                        0                       0   

      location_Hebbal Kempapura  location_Hosur Road  \
0                             0                    0   
1                             0                    0   
2                             0                    0   
3                             0                    0   
4

In [5]:
y = data['price']
X = data.drop('price',axis = 1)

In [6]:
print(y.shape)

(7120,)


In [7]:
print(X.shape)

(7120, 107)


In [8]:
X_train,X_test,y_train,y_test = train_test_split(X,y,test_size = 0.2, random_state = 0 )

In [9]:
print(X_train.shape)
print(X_test.shape)
print(y_train.shape)
print(y_test.shape)

(5696, 107)
(1424, 107)
(5696,)
(1424,)


In [10]:
math.sqrt(len(y_test))

37.73592452822641

In [11]:
model = KNeighborsRegressor(n_neighbors=5)
model.fit(X_train,y_train)

KNeighborsRegressor()

In [12]:
model_fit = model.fit(X_train,y_train)

In [13]:
round(model_fit.score(X_train,y_train),2)

0.96

In [14]:
y_pred = model_fit.predict(X_test)
results = r2_score(y_test,y_pred)
print(results)

0.9615137358075825


In [15]:
print(y_pred)

[ 68.4    35.66   59.807 ...  90.768  75.2   773.2  ]


In [16]:
print(y_test)

1189     68.0
5876     36.0
5597     61.0
3391     85.0
6122     64.0
        ...  
5830    187.0
4887     57.0
1861     90.0
3014     75.0
1670    864.0
Name: price, Length: 1424, dtype: float64


In [17]:
X_test['Actual']= y_test
X_test['y_pred']= y_pred

In [18]:
print(X_test)

      bath  balcony  total_sqft_int  bhk  price_per_sqft  \
1189   2.0      2.0          1432.0    2     4748.603352   
5876   2.0      1.0          1126.0    2     3197.158082   
5597   2.0      1.0          1200.0    2     5083.333333   
3391   2.0      1.0          1432.0    2     5935.754190   
6122   2.0      2.0          1082.0    2     5914.972274   
...    ...      ...             ...  ...             ...   
5830   4.0      2.0          2275.0    3     8219.780220   
4887   1.0      1.0           705.0    1     8085.106383   
1861   3.0      2.0          1540.0    3     5844.155844   
3014   2.0      2.0          1160.0    2     6465.517241   
1670   3.0      2.0          3875.0    3    22296.774194   

      area_typeSuper built-up  Area  area_typeBuilt-up  Area  \
1189                              1                        0   
5876                              1                        0   
5597                              0                        0   
3391                   

In [19]:
error = []
k = []
for i in range(1,21):
    model = KNeighborsRegressor(n_neighbors=i)
    model_fit = model.fit(X_train,y_train)
    err = 1 - round(model_fit.score(X_train, y_train),2)
    error.append(err)
    k.append(i)

In [20]:
pd.DataFrame({'k':k, 'error':error})


k  error
0    1   0.00
1    2   0.02
2    3   0.05
3    4   0.03
4    5   0.04
5    6   0.06
6    7   0.08
7    8   0.09
8    9   0.10
9   10   0.11
10  11   0.12
11  12   0.13
12  13   0.14
13  14   0.15
14  15   0.15
15  16   0.15
16  17   0.16
17  18   0.17
18  19   0.17
19  20   0.17

In [21]:
import joblib

In [22]:
joblib.dump(model,'houseprice_prj6.sav')

['houseprice_prj6.sav']

So, we created a model using KNN which can Predict the Price of Bangalore House. And the 96 % of accuracy is tell us that it is a pretty fair fit the model.